components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttConnectionListener.java

```diff
@@ -37,28 +36,33 @@ public MqttConnectionListener(MqttConnectionConsumer mqttConnectionConsumer) {
         this.mqttConnectionConsumer = mqttConnectionConsumer;
     }
 
-    public void onSuccess(IMqttToken token) {
-        mqttConnectionConsumer.releaseTaskSuspension();
-    }
-
-    public void onFailure(IMqttToken token, Throwable exception) {
+    public void onFailure() {
         try {
 
             int retryInterval = mqttConnectionConsumer.getMqttConnectionFactory().
                     getReconnectionInterval();
-            if (retryInterval != -1) {
-                Thread.sleep(retryInterval);
-            } else {
-                Thread.sleep(DEFAULT_RECONNECTION_INTERVAL);
-            }
+            boolean isConnected = false;
+            int retryCount = 1;
+            while (execute && !isConnected) {
+                if (retryInterval != -1) {
+                    Thread.sleep(retryInterval);
+                } else {
+                    Thread.sleep(DEFAULT_RECONNECTION_INTERVAL);
+                }
+                try {
+                    IMqttToken connectionToken = mqttConnectionConsumer.getMqttAsyncClient().connect(mqttConnectionConsumer.getConnectOptions());
+                    connectionToken.waitForCompletion();
+                    if (mqttConnectionConsumer.getMqttAsyncClient().isConnected()) {
+                        isConnected = true;
+                        log.info("Successfully reconnected MQTT inbound endpoint: " + mqttConnectionConsumer.getName());
+                    }
+                } catch (MqttException ex) {
+                    log.error("MQTT inbound endpoint " + mqttConnectionConsumer.getName()
+                            + " error while reconnecting to the broker attempt " + retryCount++);
+                }
 
-            if (execute) {
-                mqttConnectionConsumer.getMqttAsyncClient()
-                        .connect(mqttConnectionConsumer.getConnectOptions(), this);
             }
-        } catch (MqttException ex) {
-            log.error("Error while trying to subscribe to the remote", ex);
-        } catch (InterruptedException ex) {
+        }  catch (InterruptedException ex) {
```

Comment:
I think we need to handle Interrupted Exception properly. ex :-  Thread.currentThread().interrupt() 
But need to focus weather this is necessary.